### PR TITLE
Fix flakey test - TestDnsStats

### PR DESCRIPTION
**Pull request by akarpz**
### What does this PR do?

Creates a new DNS connection in each retry of the require.Eventually

### Motivation

https://datadoghq.atlassian.net/browse/NPM-3012

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/tracer/tracer_test.go [hunk 1]
```diff
@@ -1017,12 +1017,13 @@ func testDNSStats(t *testing.T, tr *Tracer, domain string, success, failure, tim
...
 
 	dnsClient := new(dns.Client)
-	dnsConn, err := dnsClient.Dial(dnsServerAddr.String())
-	require.NoError(t, err)
-	defer dnsConn.Close()
-	dnsClientAddr := dnsConn.LocalAddr().(*net.UDPAddr)
+	var dnsClientAddr *net.UDPAddr
 	require.Eventually(t, func() bool {
+		dnsConn, err := dnsClient.Dial(dnsServerAddr.String())
+		require.NoError(t, err)
```
Comment (hmahmood) [inline pkg/network/tracer/tracer_test.go:1023]: Does this fail the test? There is also a version of Eventually that passes in a context. https://pkg.go.dev/github.com/stretchr/testify/assert#EventuallyWithT

Comment (akarpz) [reply pkg/network/tracer/tracer_test.go:1023]: this passes locally, I've been running each branch 250 times to ensure it's no longer flaky:
```
--- PASS: TestTracerSuite (11.15s)
    --- PASS: TestTracerSuite/runtime_compiled (11.15s)
        --- PASS: TestTracerSuite/runtime_compiled/TestDNSStats (7.03s)
            --- PASS: TestTracerSuite/runtime_compiled/TestDNSStats/valid_domain (0.61s)
            --- PASS: TestTracerSuite/runtime_compiled/TestDNSStats/invalid_domain (0.63s)
            --- PASS: TestTracerSuite/runtime_compiled/TestDNSStats/timeout (2.61s)
        --- PASS: TestTracerSuite/runtime_compiled/TestDNSStatsWithNAT (4.11s)
PASS
ok  	github.com/DataDog/datadog-agent/pkg/network/tracer	2793.268s
```

Comment (hmahmood) [reply pkg/network/tracer/tracer_test.go]: I meant using `require.Error` there inside `require.Eventually` would fail the test if it trips?

Comment (hmahmood) [reply pkg/network/tracer/tracer_test.go]: I guess it doesn't matter since you switched to context.

**Review by hmahmood — APPROVED** [review]

File: pkg/network/tracer/tracer_test.go [hunk 2]
```diff
@@ -1069,14 +1072,22 @@ func (s *TracerSuite) TestDNSStats() {
...
 	t.Run("invalid domain", func(t *testing.T) {
 		testDNSStats(t, tr, "abcdedfg", 0, 1, 0, validDNSServer)
 	})
+}
+
+func (s *TracerSuite) TestDNSStatsWithTimeout() {
+	t := s.T()
+	cfg := testConfig()
+	cfg.CollectDNSStats = true
+	cfg.DNSTimeout = 1 * time.Second
```
Comment (hmahmood) [inline pkg/network/tracer/tracer_test.go:1089]: Can we try `0` here?

File: pkg/network/tracer/tracer_test.go [hunk 3]
```diff
@@ -1069,14 +1072,22 @@ func (s *TracerSuite) TestDNSStats() {
 	t := s.T()
 	cfg := testConfig()
 	cfg.CollectDNSStats = true
-	cfg.DNSTimeout = 1 * time.Second
+	cfg.DNSTimeout = 5 * time.Second
```
Comment (hmahmood) [inline pkg/network/tracer/tracer_test.go:1075]: We should not set this so the default gets used.

Comment (akarpz) [reply pkg/network/tracer/tracer_test.go:1075]: sure, looks like default is 15s

File: pkg/network/tracer/tracer_test.go (hunk 2 above)
Comment (akarpz) [reply pkg/network/tracer/tracer_test.go:1089]: apparently not:
`suite.go:87: test panicked: non-positive interval for NewTicker`

Comment (akarpz) [reply pkg/network/tracer/tracer_test.go:1089]: I can set it to 1ms?

File: pkg/network/tracer/tracer_test.go [hunk 4]
```diff
@@ -1017,13 +1017,16 @@ func testDNSStats(t *testing.T, tr *Tracer, domain string, success, failure, tim
...
 
 	dnsClient := new(dns.Client)
-	dnsConn, err := dnsClient.Dial(dnsServerAddr.String())
-	require.NoError(t, err)
-	defer dnsConn.Close()
-	dnsClientAddr := dnsConn.LocalAddr().(*net.UDPAddr)
-	require.Eventually(t, func() bool {
+	var dnsClientAddr *net.UDPAddr
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
+		dnsConn, err := dnsClient.Dial(dnsServerAddr.String())
```
Comment (leeavital) [inline pkg/network/tracer/tracer_test.go:1022]: This seems susceptible to the same problem I've mentioned in other DNS tests where timeouts are overcounted.

Comment (akarpz) [reply pkg/network/tracer/tracer_test.go:1022]: yea... I was thinking of removing these retries entirely and just increasing the tracer timeout for now

Comment (akarpz) [reply pkg/network/tracer/tracer_test.go:1022]: so actually we only log a timeout in the tracer when we get a response... in the failure case for this test we don't, and shouldn't see a timeout

File: pkg/network/tracer/tracer_test.go (hunk 2 above)
Comment (hmahmood) [reply pkg/network/tracer/tracer_test.go:1089]: Lets set it to `250ms`, that ticker should fire before we do the asserts and set the timeout; the test has a sleep for `500ms` so this should fire before that.

**Review by hmahmood — APPROVED** [review]

**Pull request closed without merging**